pkg/openfeature/event_handler.go

```diff
@@ -0,0 +1,243 @@
+		clientRegistry:       map[string]clientHolder{},
+		logger:               logger,
+	}
+}
+
+// clientHolder is a helper struct to hold client specific callbacks
+type clientHolder struct {
+	name   string
+	holder map[EventType][]EventCallBack
+}
+
+func newClientHolder(client string) clientHolder {
+	return clientHolder{
+		name:   client,
+		holder: map[EventType][]EventCallBack{},
+	}
+}
+
+func (e *eventHandler) updateLogger(l logr.Logger) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.logger = l
+}
+
+func (e *eventHandler) registerApiHandler(t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if e.apiRegistry[t] == nil {
+		e.apiRegistry[t] = []EventCallBack{c}
+	} else {
+		e.apiRegistry[t] = append(e.apiRegistry[t], c)
+	}
+}
+
+func (e *eventHandler) removeApiHandler(t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	entrySlice, ok := e.apiRegistry[t]
+	if !ok {
+		// nothing to remove
+		return
+	}
+
+	for i, f := range entrySlice {
+		if f == c {
+			entrySlice = append(entrySlice[:i], entrySlice[i+1:]...)
+		}
+	}
+
+	e.apiRegistry[t] = entrySlice
+}
+
+func (e *eventHandler) registerClientHandler(clientName string, t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	_, ok := e.clientRegistry[clientName]
+	if !ok {
+		e.clientRegistry[clientName] = newClientHolder(clientName)
+	}
+
+	v := e.clientRegistry[clientName]
+
+	if v.holder[t] == nil {
+		v.holder[t] = []EventCallBack{c}
+	} else {
+		v.holder[t] = append(v.holder[t], c)
+	}
+}
+
+func (e *eventHandler) removeClientHandler(name string, t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	_, ok := e.clientRegistry[name]
+	if !ok {
+		// nothing to remove
+		return
+	}
+
+	entrySlice := e.clientRegistry[name].holder[t]
+	if entrySlice == nil {
+		// nothing to remove
+		return
+	}
+
+	for i, f := range entrySlice {
+		if f == c {
+			entrySlice = append(entrySlice[:i], entrySlice[i+1:]...)
+		}
+	}
+
+	e.clientRegistry[name].holder[t] = entrySlice
+}
+
+func (e *eventHandler) registerEventingProvider(provider FeatureProvider) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	v, ok := provider.(EventHandler)
+	if !ok {
+		return
+	}
+
+	// register shutdown semaphore
+	sem := make(chan interface{})
+	e.providerShutdownHook[provider.Metadata().Name] = sem
+
+	go func() {
+		for {
+			select {
+			case event := <-v.EventChannel():
+				err := e.triggerEvent(event)
+				if err != nil {
+					e.logger.Error(err, fmt.Sprintf("error handling event type: %s", event.EventType))
+				}
+			case <-sem:
+				return
+			}
+		}
+	}()
+}
+
+func (e *eventHandler) triggerEvent(event Event) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	group, gCtx := errgroup.WithContext(ctx)
+
+	group.Go(func() error {
+		// first run API handlers
+		for _, c := range e.apiRegistry[event.EventType] {
+			f := *c
+			func() {
```

Comment:
this function and the one handling the client handlers are mostly the same, except for the client name that's being added in the latter. To improve readability of this function I suggest to make one reusable function (smth. like `e.callEventHandlerFunc`) and make use of that one for the API and Client event handlers

Reply:
Thanks, yes will refactor to extract this out.

Reply:
This is done - https://github.com/open-feature/go-sdk/pull/188/files#diff-5088e3f82f28056163763bfb774c1b8e91bb585580b4b1add90b133445b8bae6R191-R206